src/core/net/ip6.cpp

```diff
@@ -1035,8 +1036,21 @@ Error Ip6::ProcessReceiveCallback(Message &          aMessage,
             Udp::Header udp;
 
             IgnoreError(aMessage.Read(aMessage.GetOffset(), udp));
-            VerifyOrExit(Get<Udp>().ShouldUsePlatformUdp(udp.GetDestinationPort()), error = kErrorNoRoute);
-
+            switch (mIp6ReceiveFilterType)
+            {
+            case OT_IP6_RX_FILTER_MODE_FILTER_THREAD_CONTROL_TRAFFIC:
+            {
+                VerifyOrExit(Get<Udp>().ShouldUsePlatformUdp(udp.GetDestinationPort()), error = kErrorNoRoute);
+                break;
+            }
+            case OT_IP6_RX_FILTER_MODE_FILTER_ALL_THREAD_USED_PORTS:
+            {
+                VerifyOrExit(Get<Udp>().IsPortInUse(udp.GetDestinationPort()) == false, error = kErrorNoRoute);
+                break;
+            }
+            default:
+                OT_ASSERT(false);
+            }
```

Comment:
```suggestion
            VerifyOrExit(Get<Udp>().ShouldUsePlatformUdp(udp.GetDestinationPort()) && !Get<Udp>().IsPortInUse(udp.GetDestinationPort()), error = kErrorNoRoute);
```
I think it's unnecessary to define `mIp6ReceiveFilterType` after all. 
We should be able to handle UDP messages correctly without defining an extra `filter type`.
Thoughts? @MarekPorwisz @abtink @bukepo

Reply:
I had it originally that way (but without backbone sockets included in the IsPortInUse) and some tests were failing. I do not remember if all were backbone router related. I also wonder if someone has already based their design  on limited filtering and may need to filter only control traffic.

Reply:
created a separate draft PR with your proposal to see if it passes tests:
https://github.com/openthread/openthread/pull/6641

Reply:
The other PR fails on test_plat_udp_accessiblity.py: _test_srp_server. TBH I am not sure if this is related to the filters change, but that is possible.

Reply:
Very likely. Failing on `test_plat_udp_accessiblity.py` constantly would mean that the filter change does not work with platform UDP yet (need fix).